commitrmnocb/outcome.go

```diff
@@ -0,0 +1,325 @@
+	fChains := p.fChainConsensus(aggObs.FChain)
+
+	fDestChain, exists := fChains[p.cfg.DestChain]
+	if !exists {
+		return ConsensusObservation{},
+			fmt.Errorf("no consensus value for fDestChain, DestChain: %d", p.cfg.DestChain)
+	}
+
+	consensusObs := ConsensusObservation{
+		MerkleRoots: p.merkleRootConsensus(aggObs.MerkleRoots, fChains),
+		// TODO: use consensus of observed gas prices
+		GasPrices: make(map[cciptypes.ChainSelector]cciptypes.BigInt),
+		// TODO: use consensus of observed token prices
+		TokenPrices:        make(map[types.Account]cciptypes.BigInt),
+		OnRampMaxSeqNums:   p.onRampMaxSeqNumsConsensus(aggObs.OnRampMaxSeqNums, fChains),
+		OffRampNextSeqNums: p.offRampMaxSeqNumsConsensus(aggObs.OffRampNextSeqNums, fDestChain),
+		FChain:             fChains,
+	}
+
+	return consensusObs, nil
+}
+
+// Given a mapping from chains to a list of merkle roots, return a mapping from chains to a single consensus merkle
+// root. The consensus merkle root for a given chain is the merkle root with the most observations that was observed at
+// least fChain times.
+func (p *Plugin) merkleRootConsensus(
+	rootsByChain map[cciptypes.ChainSelector][]cciptypes.MerkleRootChain,
+	fChains map[cciptypes.ChainSelector]int,
+) map[cciptypes.ChainSelector]cciptypes.MerkleRootChain {
+	consensus := make(map[cciptypes.ChainSelector]cciptypes.MerkleRootChain)
+
+	for chain, roots := range rootsByChain {
+		if f, exists := fChains[chain]; exists {
+			root, count := mostFrequentElem(roots)
+
+			if count <= f {
+				// TODO: metrics
+				p.lggr.Warnf("failed to reach consensus on a merkle root for chain %d "+
+					"because no single merkle root was observed more than the expected %d times, found merkle root %d "+
+					"observed by only %d oracles, all observed merkle roots: %v",
+					chain, f, root, count, roots)
+			}
+
+			consensus[chain] = root
+		} else {
+			// TODO: metrics
+			p.lggr.Warnf("merkleRootConsensus: fChain not found for chain %d", chain)
+		}
+	}
+
+	return consensus
+}
+
+// Given a mapping from chains to a list of max seq nums on their corresponding OnRamp, return a mapping from chains
+// to a single max seq num. The consensus max seq num for a given chain is the f'th lowest max seq num if the number
+// of max seq num observations is greater or equal than 2f+1, where f is the FChain of the corresponding source chain.
+func (p *Plugin) onRampMaxSeqNumsConsensus(
+	onRampMaxSeqNumsByChain map[cciptypes.ChainSelector][]cciptypes.SeqNum,
+	fChains map[cciptypes.ChainSelector]int,
+) map[cciptypes.ChainSelector]cciptypes.SeqNum {
+	consensus := make(map[cciptypes.ChainSelector]cciptypes.SeqNum)
+
+	for chain, onRampMaxSeqNums := range onRampMaxSeqNumsByChain {
+		if f, exists := fChains[chain]; exists {
+			if len(onRampMaxSeqNums) < 2*f+1 {
+				// TODO: metrics
+				p.lggr.Warnf("could not reach consensus on onRampMaxSeqNums for chain %d "+
+					"because we did not receive more than 2f+1 observed sequence numbers, 2f+1: %d, "+
```

Comment:
```suggestion
					"because we did not receive more than 2fChain+1 (%d) observed sequence numbers, "+
```

Maybe this is clearer, matches the pattern of the log below